### PR TITLE
Add prometheus format

### LINKED ISSUE #6: Output details in prometheus format
Similarly to https://github.com/outscale-dev/osc-cost/issues/3, users may want to output data as prometheus format.

- available through `--format prometheus`

### PR DESCRIPTION
**Pull request by outscale-vbr**
This PR adds prometheus format output.

Closes  #6

### CODE REVIEW

File: src/core.rs [hunk 1]
```diff
@@ -207,6 +1043,56 @@ impl ResourceTrait for Aggregate {
...
+                .const_label("account_id", account_id)
+                .const_label("read_date_rfc3339", read_date_rfc3339)
+                .const_label("region", region)
+                .const_label("resource_id", resource_id)
+                .const_label("resource_type", resource_type);
+        let aggregate_gauge_month =
+            Gauge::with_opts(aggregate_gauge_month_opts).unwrap_or_else(|err| {
+                println!("Error aggregate gauge month: {}", err);
+                process::exit(1);
+            });
```
Comment (outscale-mdr) [inline src/core.rs:1093]: Return an error instead of exiting

Comment (outscale-mdr) [reply src/core.rs:1093]: Use `warn!`to display the message

File: src/core.rs [hunk 2]
```diff
@@ -183,8 +997,30 @@ impl error::Error for ResourceError {}
...
+        read_date_rfc3339: &String,
+        region: &String,
+        resource_id: &String,
+        resource_type: &String,
+    ) -> Result<GenericGauge<AtomicF64>, ResourceError>;
 }
 
+trait ResourceMetricsTrait {
+    fn register(&self, registry: Registry) -> Result<Registry, ResourceError>;
+    fn add(&self, price_per_hour: f64, price_per_month: f64) -> Result<(), ResourceError>;
```
Comment (outscale-mdr) [inline src/core.rs:1022]: Do we need two functions as we are only fetching one time the resource ?

File: src/core.rs [hunk 3]
```diff
@@ -163,6 +180,803 @@ impl Resources {
...
+    pub fn metrics(&self) -> Result<Registry, Box<dyn error::Error>> {
+        let mut registry = Registry::new();
+
+        for resource in &self.resources {
+            match resource {
+                Resource::Vm(vm) => {
+                    let osc_cost_version = vm
+                        .osc_cost_version
+                        .as_ref()
+                        .unwrap_or(&"v0.0.0".to_string())
```
Comment (outscale-mdr) [inline src/core.rs:192]: I don't think this is a good idea to compute with arbitrary values, if the value is not there, it should be a problem

File: src/core.rs [hunk 4]
```diff
@@ -163,6 +180,803 @@ impl Resources {
...
+        let metric_families = registry.gather();
+        encoder.encode(&metric_families, &mut buffer).unwrap();
+        Some(String::from_utf8(buffer.clone()).unwrap())
+    }
+    pub fn prometheus(&self) -> Result<String, ResourceError> {
+        let registry = match self.metrics() {
+            Ok(registry) => registry,
+            Err(e) => {
+                error!("error: cannot get registry: {}", e);
+                Registry::new()
```
Comment (outscale-mdr) [inline src/core.rs:970]: We should retur an error, no ?

File: src/core.rs [hunk 5]
```diff
@@ -183,8 +997,30 @@ impl error::Error for ResourceError {}
...
     fn price_per_hour(&self) -> Result<f32, ResourceError>;
     fn compute(&mut self) -> Result<(), ResourceError>;
+    fn gauge_hour(
+        &self,
+        osc_cost_version: &String,
+        account_id: &String,
+        read_date_rfc3339: &String,
+        region: &String,
+        resource_id: &String,
+        resource_type: &String,
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply src/core.rs]: Ok done

File: src/core.rs [hunk 6]
```diff
@@ -183,8 +997,30 @@ impl error::Error for ResourceError {}
...
+    ) -> Result<GenericGauge<AtomicF64>, ResourceError>;
+    fn gauge_month(
+        &self,
+        osc_cost_version: &String,
+        account_id: &String,
+        read_date_rfc3339: &String,
+        region: &String,
+        resource_id: &String,
+        resource_type: &String,
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply src/core.rs]: Ok done

File: src/core.rs [hunk 7]
```diff
@@ -207,6 +1043,56 @@ impl ResourceTrait for Aggregate {
...
+                .const_label("account_id", account_id)
+                .const_label("read_date_rfc3339", read_date_rfc3339)
+                .const_label("region", region)
+                .const_label("resource_id", resource_id)
+                .const_label("resource_type", resource_type);
+        let aggregate_gauge_hour =
+            Gauge::with_opts(aggregate_gauge_hour_opts).unwrap_or_else(|err| {
+                println!("Error aggregate gauge hour: {}", err);
+                process::exit(1);
+            });
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply src/core.rs:549]: Ok done

File: src/core.rs (hunk 1 above)
Comment (outscale-vbr) [reply src/core.rs:565]: Ok done

File: src/core.rs (hunk 2 above)
Comment (outscale-vbr) [reply src/core.rs]: Ok function add remove

File: src/core.rs (hunk 4 above)
Comment (outscale-vbr) [reply src/core.rs]: Replace by ?

File: src/core.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply src/core.rs:549]: Replace by ?

File: src/core.rs [hunk 8]
```diff
@@ -297,3 +655,19 @@ impl From<Resource> for Aggregate {
...
+
+pub struct AggregateMetrics {
+    aggregate_price_per_hours: GenericGauge<AtomicF64>,
+    aggregate_price_per_months: GenericGauge<AtomicF64>,
+}
+impl ResourceMetricsTrait for AggregateMetrics {
+    fn register(&self, registry: Registry) -> Result<Registry, prometheus::Error> {
+        registry
+            .register(Box::new(self.aggregate_price_per_hours.clone()))
+            .or_else(|e| Err(e))?;
```
Comment (outscale-mdr) [inline src/core.rs:667]: remove the `.or_else(|e| Err(e))`

File: src/core.rs [hunk 9]
```diff
@@ -207,6 +533,38 @@ impl ResourceTrait for Aggregate {
...
+        Ok(aggregate_gauge_hour)
+    }
+
+    fn gauge_month(&self) -> Result<GenericGauge<AtomicF64>, prometheus::Error> {
+        let aggregate_gauge_month_opts =
+            Opts::new("aggregate_price_month", "Aggregate price by month")
+                .const_label("osc_cost_version", self.osc_cost_version.as_ref().unwrap())
+                .const_label("account_id", self.account_id.as_ref().unwrap())
+                .const_label("region", self.region.as_ref().unwrap())
+                .const_label("resource_id", "aggregate".to_string())
```
Comment (outscale-mdr) [inline src/core.rs:559]: remove the `resourceId`, it is not necessary

File: src/core/load_balancers.rs [hunk 10]
```diff
@@ -25,4 +29,45 @@ impl ResourceTrait for LoadBalancer {
...
         }
     }
+    fn gauge_hour(&self) -> Result<GenericGauge<AtomicF64>, prometheus::Error> {
+        let load_balancer_gauge_hour_opts =
+            Opts::new("load_balancer_price_hour", "LoadBalancer price by hour")
+                .const_label("osc_cost_version", self.osc_cost_version.as_ref().unwrap())
+                .const_label("account_id", self.account_id.as_ref().unwrap())
+                .const_label("region", self.region.as_ref().unwrap())
+                .const_label("resource_id", self.resource_id.as_ref().unwrap())
+                .const_label("resource_type", "loadBalancer".to_string());
```
Comment (outscale-mdr) [inline src/core/load_balancers.rs:39]: We need to be consistant with the `resource_type`, maybe a constant ?

**Review by outscale-mdr — CHANGES_REQUESTED** [review]
Few things:
- Try to make the code as maintainable as possible, think when a new resource will be added
- Try not to use `unwrap` otherwise the code will panic in the middle, just let the caller function deal with the problem

**Review by jerome-jutteau — CHANGES_REQUESTED** [review]
Several things I would like to change in this PR:
- Using Opts, ... objects from Prometheus module inside a `core/resourceX.rs` file mixes things up. I would move Prometheus output implementation in a separate file using the From trait to convert a core::ResourceX to a prometheus output structure.
- I would also do that for all all output formats to separate core from the rest
- please, do not mix cargo format modifications with an other commit. It makes it hard to see what really happened in the commit. To ease things here, you can merge all commits in a single one.
- Avoid having a commit just adding a dependency with no code usage, there is no problem having this change with new code using new dependencies.

**Comment by outscale-vbr** [comment]
I change the format (create a serde), i handle your remark and i open a new PR.
https://github.com/outscale/osc-cost/pull/63

**Pull request closed without merging**